pkg/logger/gorm.go

```diff
@@ -1,95 +1,77 @@
-
-		logger := gl.logger.WithFields(Fields{
-			"sql": Fields{
-				"duration":      duration,
-				"affected_rows": affectedRows,
-				"file_location": location,
-				"values":        formattedValues,
-			},
-		})
-
-		logger.Debugf(sql)
-	} else {
-		gl.logger.Tracef("%v", values[2:]...)
-	}
+func (g gormLogger) LogMode(logger.LogLevel) logger.Interface {
+	// we ignore changes to the log level.
+	// making changes to logger level causes a change to the root logger.
+	return g
+}
+
+func (g gormLogger) Info(ctx context.Context, msg string, args ...interface{}) {
+	g.logger.Infof(msg, args...)
+}
+
+func (g gormLogger) Warn(ctx context.Context, msg string, args ...interface{}) {
+	g.logger.Warnf(msg, args...)
 }
 
-// Code blatantly stolen and modified for our needs from Gorm:
-// https://github.com/jinzhu/gorm/blob/2a3ab99/logger.go#L30-L109
-func formatValues(values []interface{}) string {
-	var formattedValues []string
-	for _, value := range values {
-		indirectValue := reflect.Indirect(reflect.ValueOf(value))
-		if indirectValue.IsValid() {
-			value = indirectValue.Interface()
-			if t, ok := value.(time.Time); ok {
-				if t.IsZero() {
-					formattedValues = append(formattedValues, fmt.Sprintf("'%v'", "0000-00-00 00:00:00"))
-				} else {
-					formattedValues = append(formattedValues, fmt.Sprintf("'%v'", t.Format("2006-01-02 15:04:05")))
-				}
-			} else if b, ok := value.([]byte); ok {
-				if str := string(b); isPrintable(str) {
-					formattedValues = append(formattedValues, fmt.Sprintf("'%v'", str))
-				} else {
-					formattedValues = append(formattedValues, "'<binary>'")
-				}
-			} else if r, ok := value.(driver.Valuer); ok {
-				if value, err := r.Value(); err == nil && value != nil {
-					formattedValues = append(formattedValues, fmt.Sprintf("'%v'", value))
-				} else {
-					formattedValues = append(formattedValues, "NULL")
-				}
-			} else {
-				switch value.(type) {
-				case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64, float32, float64, bool:
-					formattedValues = append(formattedValues, fmt.Sprintf("%v", value))
-				default:
-					formattedValues = append(formattedValues, fmt.Sprintf("'%v'", value))
-				}
-			}
-		} else {
-			formattedValues = append(formattedValues, "NULL")
-		}
+func (g gormLogger) Error(ctx context.Context, msg string, args ...interface{}) {
+	g.logger.WithError(fmt.Errorf(msg, args...)).Errorf(gormLoggerMsg)
+}
+
+func (g gormLogger) Trace(ctx context.Context, begin time.Time, fc func() (string, int64), err error) {
+	sql, rows := fc()
+
+	l := g.logger.WithFields(Fields{
+		gormLoggerTraceFieldKey: Fields{
+			"duration":      time.Since(begin),
+			"affected_rows": rows,
+			"sql":           sql,
+		},
+	})
+
+	if err == nil {
+		l.Tracef("%s", gormLoggerMsg)
+
+		return
 	}
 
-	return strings.Join(formattedValues, " ")
+	if errors.Is(err, gorm.ErrRecordNotFound) {
+		l.Tracef(err.Error())
+
+		return
+	}
+
+	l.WithError(err).Errorf(gormLoggerMsg)
```

Comment:
[q]: can we just reduce this block to one `Tracef` call instead of having those `if-statements`? Something like:

```go

sql, rows := fc()

l := g.logger.WithFields(Fields{
	gormLoggerTraceFieldKey: Fields{
		"duration":      time.Since(begin),
		"affected_rows": rows,
		"sql":           sql,
	},
})

l.WithError(err).Tracef(gormLoggerMsg)
```

Reply:
@Neurostep But then we would get the errors as trace :thinking: Is it what we want?

Reply:
Since this log will be called only in `trace` mode I think it makes sense to log these entries as `Trace` as well. Looks like in the default logger they do the same: have [error logs](https://github.com/go-gorm/gorm/blob/02b7e26f6b5dcdc49797cc44c26a255a69f3aff3/logger/logger.go#L90) dedicated for errors, and [trace logs](https://github.com/go-gorm/gorm/blob/02b7e26f6b5dcdc49797cc44c26a255a69f3aff3/logger/logger.go#L93).

Reply:
The error itself could be extracted from the execution of statement on `db` side and there is no need to handle that in here, thanks to @Neurostep for pointing that out.
I changed the code as requested, have another look please :pray: